### DISCUSSION TITLE
Sign in with Steam workaround

### DISCUSSION THREAD

**Discussion by oliv-tree** [supabase/supabase#15118] [Questions]
As shown here https://github.com/supabase/gotrue/issues/451#issuecomment-1101917396, Steam seems to be a highly demanded provider.

However, it doesn’t seem like any progress has been made on this? 

So, in the meantime I was wondering if anyone knows of a sensible method of utilising Steam’s OpenID 2.0 auth combined with Supabase.

A unique `steam_id` is returned by Steam on successful authentication. My current thought is to use this in both the `email` and `password` field when creating/logging in a user, ensuring this occurs on the server side. But would I then need to disable client-side logging in?

As far as I understand it, this should return some `id` which uniquely identifies the user inside `auth.users`, so I could add a column to my `public.users` (which uses `steam_id` as the primary key) to link these together. Particularly so the logged in user can only modify their own data in the database.

I’m currently using this library in my SvelteKit TS app to manage the auth https://github.com/LeeviHalme/node-steam-openid

**Comment by oliv-tree** [comment]
Perhaps this might help?

https://github.com/orgs/supabase/discussions/14266#discussioncomment-5866648

I’m a bit confused about refresh tokens though 

Also not sure how I would only let users access their own rows in the DB?

**Comment by oliv-tree** [comment]
There’s also this, but it uses deprecated methods 

https://github.com/orgs/supabase/discussions/1849

**Comment by oliv-tree** [comment]
What if I do 
```ts
// User logs in with Steam for the first time
// Returns their `steam_id`, then we create user as below

const { data, error } = await supabase.auth.admin.createUser({
  email: steam_id,
  password: randomPassword(),
})
// Add a trigger to add user to `public.users`
```

And then to log the user in (and what we will do in future since they have now signed up), we do something like
```ts
// Retrieve `uid` from `public.users` by their `steam_id`
const { data, error } = await supabase.auth.admin.getUserById(uid);
// Retrieve `password` from `data`?

const { data, error } = await supabase.auth.signInWithPassword({
  email: steam_id,
  password: password,
});
```

**Reply by oliv-tree** [reply]
However, it seems like maybe `supabase.auth.admin.getUserById` needs to be on the server, but `supabase.auth.signInWithPassword` needs to be on the client?

**Reply by oliv-tree** [reply]
Actually is `supabase.auth.admin.createUser()` creating an admin user? I'm confused

**Reply by Revadike** [reply]
> However, it seems like maybe `supabase.auth.admin.getUserById` needs to be on the server, but `supabase.auth.signInWithPassword` needs to be on the client?

You can avoid a password altogether by generating a OTP link and make the user redirect to it. 
https://supabase.com/docs/reference/javascript/auth-admin-generatelink?example=generate-a-magic-link

Let me know if this works for you, or if you found a better solution yourself in the meantime.

**Comment by Revadike** [comment]
[I attempted a PR for Steam provider support](https://github.com/supabase/auth/pull/1919)
However, I think the main problem is Supabase's reliability on user emails, so I am not sure if this will (fully) work.
